File: sdk-js/src/exceptions.ts
/**
 * Exception classes for Universal Agent Connector SDK
 */

export class UniversalAgentConnectorError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'UniversalAgentConnectorError';
    Object.setPrototypeOf(this, UniversalAgentConnectorError.prototype);
  }
}

export class APIError extends UniversalAgentConnectorError {
  public readonly statusCode?: number;
  public readonly response?: any;

  constructor(message: string, statusCode?: number, response?: any) {
    super(message);
    this.name = 'APIError';
    this.statusCode = statusCode;
    this.response = response;
    Object.setPrototypeOf(this, APIError.prototype);
  }
}

export class AuthenticationError extends APIError {
  constructor(message: string = 'Authentication failed. Check your API key.', statusCode: number = 401, response?: any) {
    super(message, statusCode, response);
    this.name = 'AuthenticationError';
    Object.setPrototypeOf(this, AuthenticationError.prototype);
  }
}

export class NotFoundError extends APIError {
  constructor(message: string = 'Resource not found', statusCode: number = 404, response?: any) {
    super(message, statusCode, response);
    this.name = 'NotFoundError';
    Object.setPrototypeOf(this, NotFoundError.prototype);
  }
}

export class ValidationError extends APIError {
  constructor(message: string = 'Validation error', statusCode: number = 400, response?: any) {
    super(message, statusCode, response);
    this.name = 'ValidationError';
    Object.setPrototypeOf(this, ValidationError.prototype);
  }
}

export class RateLimitError extends APIError {
  constructor(message: string = 'Rate limit exceeded', statusCode: number = 429, response?: any) {
    super(message, statusCode, response);
    this.name = 'RateLimitError';
    Object.setPrototypeOf(this, RateLimitError.prototype);
  }
}

export class ConnectionError extends UniversalAgentConnectorError {
  constructor(message: string) {
    super(message);
    this.name = 'ConnectionError';
    Object.setPrototypeOf(this, ConnectionError.prototype);
  }
}

File: sdk/universal_agent_connector/exceptions.py
"""
Exception classes for the Universal Agent Connector SDK
"""


class UniversalAgentConnectorError(Exception):
    """Base exception for all SDK errors"""
    pass


class APIError(UniversalAgentConnectorError):
    """Raised when API returns an error response"""
    
    def __init__(self, message: str, status_code: int = None, response: dict = None):
        super().__init__(message)
        self.status_code = status_code
        self.response = response or {}


class AuthenticationError(APIError):
    """Raised when authentication fails"""
    pass


class NotFoundError(APIError):
    """Raised when a resource is not found (404)"""
    pass


class ValidationError(APIError):
    """Raised when request validation fails (400)"""
    pass


class RateLimitError(APIError):
    """Raised when rate limit is exceeded (429)"""
    pass


class ConnectionError(UniversalAgentConnectorError):
    """Raised when connection to API fails"""
    pass
